### PR TITLE
WIP: Use Troika directives in TroikaHost (#31)

### PR DESCRIPTION
**Pull request by oiffrig**
(#31) Converts the Slurm directives in TroikaHost to Troika directives

WIP: directives kept as plain Slurm for now:
* `distribution`, `reservation` are supported from Troika 0.2.2 only, support disabled for now
* `tmpdir`, `sthost` are ECMWF-specific, and should be in the local Troika configuration rather than here
* `hint` is not supported by Troika in its most general form, is it needed past `--[no]multithread`?

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/pyflow?pullRequest=34) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/pyflow?pullRequest=34) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/pyflow?pullRequest=34) it.</sub>

File: pyflow/host.py [hunk 1]
```diff
@@ -1014,35 +1014,70 @@ def script_submit_arguments(self, submit_arguments):
...
+            "error": "error",
+            "priority": "priority",
+            "tmpdir": "tmpdir_size",
+            "enable_hyperthreading": "enable_hyperthreading",
+            "distribution": "distribution",  # XXX: troika >=0.2.2
+            "reservation": "reservation",  # XXX: troika >=0.2.2
+        }
+
+        slurm_resources = {
+            "tmpdir": "--gres=ssdtmp:",  # XXX: ECMWF-specific
```
Comment (corentincarton) [inline pyflow/host.py:1042]: Shouldn't this be in the troika config? We would then rely here on the troika tmdir_size definition instead?

File: pyflow/host.py [hunk 2]
```diff
@@ -1014,35 +1014,70 @@ def script_submit_arguments(self, submit_arguments):
         """
         Accepted submit arguments:
         """
+
+        TROIKA_0_2_2 = False
```
[the remark this answers is not in the document]
Comment (oiffrig) [reply pyflow/host.py:1018]: Yes, that was the idea. Not sure how best to implement it, to be honest. The nicest would be to wait till Troika 0.2.2 is widely adopted, but if we want to go forward we will need a way for the user to set this flag indeed.

File: pyflow/host.py (hunk 1 above)
Comment (oiffrig) [reply pyflow/host.py:1042]: It should. I left it here for backwards compatibility, but I would very much prefer using the Troika directive with an overload in the local config.

**Comment by corentincarton** [comment]
@oiffrig, I just had a chat with some colleagues from production who are running on LUMI and they need new options like "#SBATCH --partition=standard-g" and "#SBATCH --exclusive". At the moment, we always need to add manually the new options in pyflow, which is not very flexible. Couldn't we have a special key in the submit arguments (user_defined for instance) that would give a list of directives. For instance:
```
execution_contexts:
    parallel:
        nodes: 40
        tasks_per_node: 8 
        user_defined:
             - "#SBATCH --exclusive"
             - "#SBATCH --partition=standard-g"
             - "#TROIKA --my_troika_option=values"
```
It's the first design that comes out of my mind so happy to discuss if you have a better idea!

**Comment by oiffrig** [comment]
Could be an option, I don't have an issue with that. What I'm doing in Troika is allowing to configure a mapping that extends/overrides the set of directives. We could have something similar in pyflow, although such a mapping is very much host-dependent.

**Comment by oiffrig** [comment]
Also, Troika supports `#TROIKA partition=foo` already, I just need to add it to the TroikaHost

**Comment by corentincarton** [comment]
That's a good point.. maybe we could extend the mapping in the constructor. We could then have a special configuration in wellies for atos or lumi for instance.

**Comment by oiffrig** [comment]
I was wondering, do you think it would be useful to pass all submit arguments to Troika, keeping the overrides where necessary? That way, if somebody wants to use a new Troika directive, they will not be limited by pyflow supporting it. We can easily keep the translation where needed.

**Comment by corentincarton** [comment]
I'm just wondering if we shouldn't just drop the translation and throw an error to force people to use the troika directives instead... It is the TroikaHost after all!

**Pull request merged**